README.md
# sus-io

A SUS (Sliding Universal Score) parser and generator.

[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)

## Functionality
- Parse sus into tick-based objects.
- Allow json output.

## Usage

### ``sus.loads(data: str)``
```python
import sus

print(sus.loads("#00002: 4\n#BPM01: 120\n#00008: 01"))
```

### ``sus.load(fp: TextIO)``
```python
import sus

with open("score.sus", "r") as f:
    score = sus.load(f)
    print(score)
```

### ``Score(...).to_json(...)``, ``Score.from_json(...)``
```python
import sus
from sus import Score

with open("score.sus", "r") as fi, open("score.json", "w") as fo:
    score = sus.load(fi)
    json = score.to_json(indent=4)
    fo.write(json)

    print(Score.from_json(json))
```

### ``sus.dump(score: Score)``, ``sus.dumps(score: Score)``
```python
import sus

with open("score.sus", 'r') as fi, open(sus_file.with_suffix('.dumped.sus'), 'w') as fd:
    score = sus.load(fi)
    print(sus.dumps(score))
    sus.dump(score, fd, comment='Custom comment.', space=False)
```

## Todo

- Acknowledgement
- Add example I/O
- Contribution Guide
- High Speed
- etc.

## License

MIT © 2021 mkpoli


CHANGELOG.md
# Change Log

## [0.2.0] - 2021-01-03

### Added

- Added sus generator.
- Exported ``Score``, ``Metadata``, and ``Note`` class.

### Fixed

- Fixed tick is not int.

## [0.1.1] - 2021-01-03

### Fixed

- Fixed not returning in ``sus.load()`` resulting in None output.

## [0.1.0] - 2022-01-02

### Added

- Initialized Project.
- Added sus parser.


tests/test.py
# from analyzer import analyze
import sus
from pathlib import Path
def main():
    for sus_file in (Path('.') / 'tests').glob('*.sus'):
        if sus_file.name.endswith('.d.sus'):
            continue

        print('----------------------------------------------------------------')
        print(f'  {sus_file.name}')
        print(f'---------------------------------------------------------------')

        with (
            open(sus_file, 'r') as fi,
            open(sus_file.with_suffix('.json'), 'w') as fo,
            open(sus_file.with_suffix('.d.sus'), 'w') as fd
        ):
            score = sus.load(fi)
            fo.write(score.to_json(indent=4))
            
            print(sus.Score.from_json(score.to_json(indent=4)))
            
            sus.dump(score, fd, comment='Custom comment.', space=True)
        print('----------------------------------------------------------------')
        print()
        print()

if __name__ == '__main__':
    main()


src/sus/loader.py
import logging
import re

from collections import defaultdict
from typing import Callable, TextIO
from .schemas import Score, Note, Metadata

logger = logging.getLogger(__name__)

def process_metadata(lines: list[tuple[str]]) -> Metadata:
    result = {}
    for line in lines:
        if len(line) == 2:
            key, value = line
        else:
            key = line[0]
            value = None
        key = key[1:]
        value = value.strip('"') if value != None else None
        if key == 'TITLE':
            result['title'] = value
        elif key == 'SUBTITLE':
            result['subtitle'] = value
        elif key == 'ARTIST':
            result['artist'] = value
        elif key == 'GENRE':
            result['genre'] = value
        elif key == 'DESIGNER':
            result['designer'] = value
        elif key == 'DIFFICULTY':
            result['difficulty'] = value
        elif key == 'PLAYLEVEL':
            result['playlevel'] = value
        elif key == 'SONGID':
            result['songid'] = value
        elif key == 'WAVE':
            result['wave'] = value
        elif key == 'WAVEOFFSET':
            result['waveoffset'] = float(value)
        elif key == 'JACKET':
            result['jacket'] = value
        elif key == 'BACKGROUND':
            result['background'] = value
        elif key == 'MOVIE':
            result['movie'] = value
        elif key == 'MOVIEOFFSET':
            result['movieoffset'] = float(value)
        elif key == 'BASEBPM':
            result['basebpm'] = float(value)
        elif key == 'REQUEST':
            if 'requests' not in result:
                result['requests'] = []
            result['requests'].append(value)
    return Metadata.from_dict(result)

def process_score(lines: list[tuple[str]], metadata: list[tuple[str]]) -> Score:
    processed_metadata = process_metadata(metadata)

    try:
        ticks_per_beat_request = [int(request.split()[1]) for request in processed_metadata.requests if request.startswith('ticks_per_beat')] if processed_metadata.requests else []
        ticks_per_beat = ticks_per_beat_request[0]
    except IndexError:
        logger.warning('No ticks_per_beat request found, defaulting to 480.')
        ticks_per_beat = 480
    
    bar_lengths: list[tuple[int, float]] = []    
    for header, data in lines:
        if len(header) == 5 and header.endswith('02') and header.isdigit():
            bar_lengths.append((int(header[0:3]), float(data)))
    
    sorted_bar_lengths = sorted(bar_lengths, key=lambda x: x[0])

    ticks = 0
    
    bars = list(reversed(
        [
            (
                measure, int(beats * ticks_per_beat),
                ticks := ticks +
                    int((measure - sorted_bar_lengths[i - 1][0]) * sorted_bar_lengths[i - 1][1] * ticks_per_beat if i > 0 else 0)
            ) for i, (measure, beats) in enumerate(sorted_bar_lengths)
        ]
    ))
    
    def to_tick(measure: int, i: int, total: int) -> int:
        bar = next(bar for bar in bars if measure >= bar[0])
        if not bar: raise ValueError(f'Measure {measure} is out of range.')
        (bar_measure, ticks_per_measure, ticks) = bar
        
        return ticks + (measure - bar_measure) * ticks_per_measure + (i * ticks_per_measure) // total

    bpm_map = {}
    bpm_change_objects = []
    tap_notes = []
    directional_notes = []
    streams = defaultdict(list)

    for header, data in lines:
        if (len(header) == 5 and header.startswith('BPM')):
            bpm_map[header[3:]] = float(data)
        elif (len(header) == 5 and header.endswith('08')):
            bpm_change_objects += to_raw_objects(header, data, to_tick)
        elif (len(header) == 5 and header[3] == '1'):
            tap_notes += to_note_objects(header, data, to_tick)
        elif (len(header) == 6 and header[3] == '3'):
            channel = header[5]
            streams[channel] += to_note_objects(header, data, to_tick)
        elif (len(header) == 5 and header[3] == '5'):
            directional_notes += to_note_objects(header, data, to_tick)
            

    slide_notes = []
    for stream in streams.values():
        slide_notes += to_slides(stream)
    
    bpms = [
        (tick, bpm_map[value] or 0)
        for tick, value in
        sorted(bpm_change_objects, key=lambda x: x[0])
    ]
    
    return Score(
        metadata=processed_metadata,
        taps=tap_notes,
        directionals=directional_notes,
        slides=slide_notes,
        bpms=bpms,
        bar_lengths=bar_lengths
    )
    
def to_slides(stream: list[Note]) -> list[list[Note]]:
    slides: list[list[Note]] = []  
    current: list[Note] = None
    for note in sorted(stream, key=lambda x: x.tick):
        if not current:
            current = []
            slides.append(current)

        current.append(note)
        
        if note.type == 2:
            current = None
    return slides

def to_note_objects(header: int, data: str, to_tick: Callable[[int, int, int], int]) -> list[Note]:
    return [
        Note(
            tick=tick,
            lane=int(header[4], 36),
            width=int(value[1], 36),
            type=int(value[0], 36),
        )
        for tick, value in to_raw_objects(header, data, to_tick)
    ]

def to_raw_objects(header: int, data: str, to_tick: Callable[[int, int, int], int]) -> list[tuple[int, str]]:
    measure = int(header[:3])
    values = list(enumerate(re.findall(r'.{2}', data)))
    return [
        (to_tick(measure, i, len(values)), value)
        for i, value in values
        if value != '00'
    ]

def load(fp: TextIO) -> Score:
    return loads(fp.read())

def loads(data: str) -> Score:
    """
    Parse SUS data into a Score object.

    :param data: The score data.
    :return: A Score object.
    """
    metadata = []
    scoredata = []
    for line in data.splitlines():
        if not line.startswith('#'):
            continue
        line = line.strip()
        match = re.match(r'^#(\w+):\s*(.*)$', line)
        if match:
            scoredata.append(match.groups())
        else:
            metadata.append(tuple(line.split(' ', 1)))

    return process_score(scoredata, metadata)

src/sus/schemas.py
from dataclasses import dataclass, field
from dataclasses_json import dataclass_json, LetterCase, Undefined
from typing import Optional

from dataclasses_json.cfg import config

def exclude_none():
    return field(default=None, metadata=config(exclude=lambda x: x is None))

@dataclass_json(undefined=Undefined.EXCLUDE)
@dataclass
class Metadata:
    title: Optional[str] = exclude_none()
    subtitle: Optional[str] = exclude_none()
    artist: Optional[str] = exclude_none()
    genre: Optional[str] = exclude_none()
    designer: Optional[str] = exclude_none()
    difficulty: Optional[str] = exclude_none()
    playlevel: Optional[str] = exclude_none()
    songid: Optional[str] = exclude_none()
    wave: Optional[str] = exclude_none()
    waveoffset: Optional[float] = exclude_none()
    jacket: Optional[str] = exclude_none()
    background: Optional[str] = exclude_none()
    movie: Optional[str] = exclude_none()
    movieoffset: Optional[float] = exclude_none()
    basebpm: Optional[float] = exclude_none()
    requests: Optional[list[str]] = exclude_none()

@dataclass_json(letter_case=LetterCase.CAMEL)
@dataclass
class Note:
    tick: int
    lane: int
    width: int
    type: int

@dataclass_json(letter_case=LetterCase.CAMEL)
@dataclass
class Score:
    metadata: Metadata
    taps: list[Note]
    directionals: list[Note]
    slides: list[list[Note]]
    bpms: list[tuple[int, float]]
    bar_lengths: list[tuple[int, float]]
    
@dataclass
class BarLength:
    start_tick: int
    measure: int
    value: float

src/sus/dumper.py
import base36
import math

from . import __version__
from .schemas import Score, BarLength

from collections import defaultdict
from typing import Optional, TextIO, Union
from dataclasses import fields

class ChannelProvider:
    channel_map: dict[int, tuple[int, int]]
    
    def __init__(self):
        self.channel_map = { key: (0, 0) for key in range(36) }
    
    def generate_channel(self, start_tick: int, end_tick: int) -> int:
        for key, (start, end) in self.channel_map.items():
            if (start == 0 and end == 0) or end_tick < start or end < start_tick:
                self.channel_map[key] = (start_tick, end_tick)
                return key
        raise Exception('No more channel available.')

def dump(score: Score, fp: TextIO, **kw) -> None:
    """
    Dump a Score object into a SUS file.
    
    :param score: The score object to dump.
    :param score: The score object to dump.
    :param space: Whether to add a space after the tag (with space: "#00010: 00", without: "#00010:00").
    """
    fp.write(dumps(score, **kw))

def format_number(value: float) -> str:
    """
    Format a number into a string, where “.0” is removed if number does not have a decimal part.

    :param value: The number to format.
    """
    return str(int(value) if value % 1 == 0 else value)

def format_value(value: Union[str, float], is_str: bool) -> str:
    return f'"{value}"' if is_str else format_number(value)

def dumps(
    score: Score,
    comment: str=f'This file was generated by sus-io v{__version__} (Python).',
    space = False
) -> str:
    """
    Dump a Score object into a string in SUS format.
    
    :param score: The score object to dump.
    :param space: Whether to add a space after the tag (with space: "#00010: 00", without: "#00010:00").
    :return: SUS data as a string.
    """
    lines = []
    
    # Metadata
    lines.append(comment)
    
    ticks_per_beat = 480
    for field in fields(score.metadata):
        attr = getattr(score.metadata, field.name)
        if attr is None:
            continue
        if field.name != 'requests':
            lines.append(f'#{field.name.upper()} {format_value(attr, field.type is Optional[str])}')
        else:
            lines.append('')
            for request in score.metadata.requests:
                lines.append(f'#REQUEST "{request}"')
                if request.startswith('ticks_per_beat'):
                    ticks_per_beat = int(request.split()[1])
    lines.append('')
    
    # Scoredata
    note_maps = defaultdict(lambda: { 'raws': [], 'ticks_per_measure': 0 })
    
    bar_lengths = sorted(score.bar_lengths, key=lambda x: x[0])
    bpms = sorted(score.bpms, key=lambda x: x[0])
    taps = sorted(score.taps, key=lambda note: note.tick)
    directionals = sorted(score.directionals, key=lambda note: note.tick)
    slides = sorted(score.slides, key=lambda x: x[0].tick)

    for measure, value in bar_lengths:
        lines.append(f'#{measure:03}02:{" " if space else ""}{format_number(value)}')
    lines.append('')

    accumulated_ticks = 0
    
    bar_lengths_in_ticks = []
    
    for index, (measure, value) in enumerate(bar_lengths):
        nextMeasure = bar_lengths[index + 1][0] if index + 1 < len(bar_lengths) else 0
        start_tick = accumulated_ticks
        accumulated_ticks += int((nextMeasure - measure) * value * ticks_per_beat)
        bar_lengths_in_ticks.append(BarLength(start_tick, measure, value))
    
    bar_lengths_in_ticks.reverse()
    
    def push_raw(tick: int, info: str, data: str):
        for bar_length in bar_lengths_in_ticks:
            if tick >= bar_length.start_tick:
                current_measure = bar_length.measure + int((tick - bar_length.start_tick) / ticks_per_beat / bar_length.value)
                note_map = note_maps[f'{current_measure:03}{info}']
                note_map['raws'].append([tick - bar_length.start_tick, data])
                note_map['ticks_per_measure'] = int(bar_length.value * ticks_per_beat)
                break
    
    if len(bpms) >= 36 ** 2 - 1:
        raise Exception(f'Too much BPMS ({bpms.length} >= 36^2 -1 = {36 ** 2 - 1})')

    bpm_identifiers = {}
    for tick, value in bpms:
        identifier = base36.dumps(len(bpm_identifiers) + 1).zfill(2)
        if value not in bpm_identifiers:
            bpm_identifiers[value] = identifier
            lines.append(f'#BPM{bpm_identifiers[value]}:{" " if space else ""}{format_number(value)}')
        push_raw(tick, '08', bpm_identifiers[value])
    lines.append('')

    for note in taps:
        push_raw(note.tick, f'1{base36.dumps(note.lane)}', f'{note.type}{base36.dumps(note.width)}')

    for note in directionals:
        push_raw(note.tick, f'5{base36.dumps(note.lane)}', f'{note.type}{base36.dumps(note.width)}')

    provider = ChannelProvider()
    for steps in slides:
        start_tick = steps[0].tick
        end_tick = steps[-1].tick
        channel = provider.generate_channel(start_tick, end_tick)
        for note in steps:
            push_raw(note.tick, f'3{base36.dumps(note.lane)}{base36.dumps(channel)}', f'{note.type}{base36.dumps(note.width)}')
    
    for tag, note_map in note_maps.items():
        gcd = note_map['ticks_per_measure']
        for raw in note_map['raws']:
            gcd = math.gcd(raw[0], gcd)
        data = {}
        for raw in note_map['raws']:
            data[(raw[0] % note_map['ticks_per_measure'])] = raw[1]
        values = []
        for i in range(0, note_map['ticks_per_measure'], gcd):
            values.append(data.get(i) or '00')
        lines.append(f'#{tag}:{" " if space else ""}{"".join(values)}')

    lines.append('')

    return '\n'.join(lines)
